File: StatusBarSpeedometer/app/src/main/java/ch/rmy/android/statusbar_tacho/utils/EventSource.java
package ch.rmy.android.statusbar_tacho.utils;

import java.util.HashSet;
import java.util.Set;

public class EventSource<T> {

    private final Set<Observer<T>> observers = new HashSet<>();

    public Destroyable add(final Observer<T> observer) {
        observers.add(observer);
        return new Destroyable() {
            @Override
            public void destroy() {
                observers.remove(observer);
            }
        };
    }

    public Destroyable bind(final Observer<T> observer, T initialValue) {
        observer.on(initialValue);
        return add(observer);
    }

    public void notify(T item) {
        for (Observer observer : observers) {
            observer.on(item);
        }
    }

    public interface Observer<T> {

        void on(T item);

    }

}

File: StatusBarSpeedometer/app/src/main/java/ch/rmy/android/statusbar_tacho/utils/EventSource.kt
package ch.rmy.android.statusbar_tacho.utils

import java.util.*

class EventSource<T> {

    private val observers = HashSet<Observer<T>>()

    fun add(observer: Observer<T>): Destroyable {
        observers.add(observer)
        return object : Destroyable {
            override fun destroy() {
                observers.remove(observer)
            }
        }
    }

    fun bind(observer: Observer<T>, initialValue: T): Destroyable {
        observer.on(initialValue)
        return add(observer)
    }

    fun notify(item: T) {
        for (observer in observers) {
            observer.on(item)
        }
    }

    interface Observer<T> {

        fun on(item: T)

    }

}
